### PR TITLE
Make keyboard activation consistent

### LINKED ISSUE #1084: Accessibility issue: inconsistent and incorrect keyboard activation of elements
**UV version:** _universalviewer@4.0.25_

**I'm submitting a:** *bug report => please fork one of these [codesandbox examples](https://github.com/UniversalViewer/universalviewer/wiki/UV-Examples) with a repro of your issue and include a link to it below*

### Issue description
When navigating by keyboard, it is very inconsisent how elements are activated: space or return key. Most work with both, some only with space, others only with return. This makes keyboard navigation confusing. 
- examples for return key only: go, +, -, rotate, items inside the share modal and inside the more information sidebar 
- examples for space key only: image controls forward and backward and also opening and closing the sidebars, switching between link and embed options in the share modal 

On top of that, activating items by spacebar will also scroll the viewport to the bottom of the page. This will happen even for items that require the spacebar to activate (for example next-image-button in image controls will go to the next image but also scroll to the bottom of the page which means the image may become completely not visible).

### Steps to reproduce
1. open [this manifest](https://www.universalviewer.dev/#?xywh=583%2C1062%2C2492%2C1379&iiifManifestId=https%3A%2F%2Feap.bl.uk%2Farchive-file%2FEAP500-1-1%2Fmanifest&cv=)
2. navigate the page and try out activating controls with space and return key  
3. particularly, navigate to the image viewer area and inside there to the next button
4. activate by space bar and observe the viewport moving to the bottom of the page 

### Expected behaviour
Consistent keys used for activating elements. Commonly, it's return for buttons and links and space for activating options in forms and the like. But having both for all elements would also work

The space key should never scroll the user anywhere but especially not if it is used to activate links in other parts of the viewer.

### Possible fix
It would probably help a lot if the image controls were to be turned into semantic html rather than divs, because buttons come with built-in functionality like that. 

### WCAG criterion
[3.2.4 Consistent Identification (Level AA)](https://www.w3.org/WAI/WCAG22/Understanding/consistent-identification.html)

**EDIT**:

MDN suggests the following behaviour for links and buttons: 
- links are activated by ENTER
- buttons are activated by SPACE or ENTER
and to make sure that buttons behave like buttons and links like links to be consistent.
([source](https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Roles/button_role#accessibility))

### LINKED ISSUE #1157: Index tab does not open when user uses tab>Index and keyboard enter button.
**UV version:**

```
 universalviewer@4.025
```

**I'm submitting a:**

- [X ] bug report => please fork one of these [codesandbox examples](https://github.com/UniversalViewer/universalviewer/wiki/UV-Examples) with a repro of your issue and include a link to it below
- [ ] feature request => please use the [user stories repo](https://github.com/UniversalViewer/user-stories)
- [ ] support request => Please do not submit support requests here, use [stackoverflow](http://stackoverflow.com/questions/tagged/iiif)

**Current behavior:**
Using the keyboard the user can tab through the various buttons, on getting to the Index tab in the centre panel with the button active clicking the keyboard enter button the index tab is NOT opened

**Expected behavior:**
Using the the keyboard, tab to the index tab, ensure the index tab is active and when the keyboard enter button is hit the Index tab should open

**Steps to reproduce:**
1. Tab to the Index tab using the keyboard
2. The index tab is active - surrounding white border is visible
3. Hit the enter button on the keyboard
4. The index tab is not opened
**Note the same is true for the Thumbnail tab**

**Other information:**
Keyboard Accessibility Issue
image

### PR DESCRIPTION
**Pull request by LlGC-jop**
Fixes #1084 
Fixes #1157 

All buttons and pseudo-buttons that previously weren't able to be activated via enter/spacebar should now be able to be.

I've modified Panel.ts:onAccessibleClick() to check whether the key is either enter or space and in the case of the latter, prevent the default action (scrolling the page). 

It also now has a fourth arg to specify whether an anchor should be treated as a button i.e. spacebar should activate it. This is for things like tabs where they're an anchor but work like a button.

Some elements that were `<div>` or `<a>` have been changed to `<button>`.

I've also made the thumbnails in the left panel work with space+enter, however the gallery view wasn't possible. For some reason you can't even focus inside it, but as it's the iiif-gallery-component I've not looked into it.

### CODE REVIEW

**Comment by demiankatz** [comment]
@LlGC-jop, the gallery portion of the problem should be addressed by https://github.com/IIIF-Commons/iiif-gallery-component/pull/20 -- if you have time to try this out, I'd welcome a review so we can get it merged and released so we can incorporate it into UV.

There is work in progress on the left panel thumbnail activation in #1155. In that implementation, spacebar does not activate the thumbs, because it instead pages down through the thumbnail list. I can potentially see some value in that approach, but I'm not really sure which is better. In any case, we'll need to reconcile the work in #1155 with the work here. Do you mind taking a look at that PR? If that solution is acceptable, it might be best to merge that first (to get a smaller, more granular commit) and then resolve conflicts over here. If there are problems, maybe it would be easier to close that one and focus on the version here.

File: src/content-handlers/iiif/modules/uv-shared-module/BaseExpandPanel.ts
Comment (demiankatz) [inline src/content-handlers/iiif/modules/uv-shared-module/BaseExpandPanel.ts:68]: Do you need to set anchor as button here since above you have changed the anchor to a button?

File: src/content-handlers/iiif/modules/uv-pdfheaderpanel-module/PDFHeaderPanel.ts
Comment (demiankatz) [inline src/content-handlers/iiif/modules/uv-pdfheaderpanel-module/PDFHeaderPanel.ts:143]: Same question here...

File: src/content-handlers/iiif/modules/uv-pdfcenterpanel-module/PDFCenterPanel.ts
Comment (demiankatz) [inline src/content-handlers/iiif/modules/uv-pdfcenterpanel-module/PDFCenterPanel.ts:198]: Maybe I'm missing something, but why do you need the anchor as button setting in two places in this file, when these are buttons now and were divs in the past?

File: src/content-handlers/iiif/modules/uv-pagingheaderpanel-module/PagingHeaderPanel.ts
Comment (demiankatz) [inline src/content-handlers/iiif/modules/uv-pagingheaderpanel-module/PagingHeaderPanel.ts:416]: Another anchor that has changed to a button...

File: src/content-handlers/iiif/modules/uv-openseadragoncenterpanel-module/OpenSeadragonCenterPanel.ts
Comment (demiankatz) [inline src/content-handlers/iiif/modules/uv-openseadragoncenterpanel-module/OpenSeadragonCenterPanel.ts:371]: The buttons here all appear to be `div` elements; do we need the "anchor as button" setting?

**Review by demiankatz — CHANGES_REQUESTED** [review]
Thanks, @LlGC-jop. I have a few comments below, but they're really all the same comment: there seem to be a few places where you are setting the new "anchor as button" argument to true, but you don't need to, because those elements aren't anchors. Maybe this is intentional, because you want to future-proof in case the element types are changed later... in which case I'd rename `anchorAsButton` to `treatAsButton` or something more generic to prevent possible future confusion. However, I think my preference would be not to pass extra arguments unless they're necessary, since having sequences of unlabeled booleans generally makes the code less readable. :-)

My apologies if this is all truly necessary and I'm just misunderstanding something!

**Comment by LlGC-jop** [comment]
@demiankatz quite right, I'd initially done the code to make the anchors work, and then decided to try changing the elements to the correct type, just to see if there were any knock-on styling issues.

There weren't, but I didn't go back and change the function calls, which I'll do this afternoon.

re #1155, the change I made was to the react component and should ensure they're only activated on space or enter. I'll take a look at that PR in a bit and see what's changed.

As for the gallery, I'll fork it and see how it works - my plan was a tabindex of -1 as well :)

**Comment by demiankatz** [comment]
Thanks, @LlGC-jop, this all sounds good to me -- let me know if/when I can be of further assistance. :-)

File: src/content-handlers/iiif/modules/uv-pdfheaderpanel-module/PDFHeaderPanel.ts
Comment (LlGC-jop) [reply src/content-handlers/iiif/modules/uv-pdfheaderpanel-module/PDFHeaderPanel.ts]: This should be resolved now

File: src/content-handlers/iiif/modules/uv-pdfcenterpanel-module/PDFCenterPanel.ts
Comment (LlGC-jop) [reply src/content-handlers/iiif/modules/uv-pdfcenterpanel-module/PDFCenterPanel.ts]: This should be resolved now

File: src/content-handlers/iiif/modules/uv-pagingheaderpanel-module/PagingHeaderPanel.ts
Comment (LlGC-jop) [reply src/content-handlers/iiif/modules/uv-pagingheaderpanel-module/PagingHeaderPanel.ts]: This should be resolved now

File: src/content-handlers/iiif/modules/uv-openseadragoncenterpanel-module/OpenSeadragonCenterPanel.ts
Comment (LlGC-jop) [reply src/content-handlers/iiif/modules/uv-openseadragoncenterpanel-module/OpenSeadragonCenterPanel.ts]: No they don't. Not sure why I did that here. Have removed.

File: src/content-handlers/iiif/modules/uv-shared-module/BaseExpandPanel.ts
Comment (LlGC-jop) [reply src/content-handlers/iiif/modules/uv-shared-module/BaseExpandPanel.ts]: Not anymore. Have updated to remove args 3 and 4 in this file and others below.

**Comment by LlGC-jop** [comment]
Hi @demiankatz ,

I've made the changes necessary, and left a comment on #1155 about what to do with the thumbnail code.

**Review by demiankatz — COMMENTED** [review]
Thanks for the latest chhanges, @LlGC-jop -- looks like all of my code feedback has been addressed, and the diffs are now a bit easier to read!

We shouldn't merge this until testers have had a chance to try everything and we've finished discussing the other PRs that interrelate with this, so I'll just make this review a "comment" for now... but I'd certainly encourage testing to go forward; I don't expect too much more to change here after this point.

**Comment by LlGC-jop** [comment]
I've merged dev into this PR and resolved the conflicts, keep the changes from #1155

**Review by demiankatz — APPROVED** [review]
Thanks again! From my perspective, this is ready to go as soon as it gets the thumbs-up from testers. We can deal with the issues related to the iiif-gallery-component separately, since that work stands alone.

**Comment by scoutb-cogapp** [comment]
I have re-tested this. It fixes #1157. It fixes #1084 also. But I agree with @LlGC-jop in this issue on which he commented in the related issue #1155: 

> I suppose the question is do we treat thumbnails as buttons or links? I would argue that as they respond to interaction to change the view without taking the user anywhere else that they're essentially a button, just one with an image rather than text.
> 
> _Originally posted by @LlGC-jop in https://github.com/UniversalViewer/universalviewer/issues/1155#issuecomment-2432160271_

I agree that these are buttons rather than links and ideally should respond to spacebar as well as return key.

**Comment by demiankatz** [comment]
Thanks, @scoutb-cogapp, I've opened #1176 to revise the thumbnail keyboard behavior as you suggest. If somebody wants to review that, I can merge it once it's approved -- though if @LlGC-jop would prefer to incorporate the fix here in a more elegant way, I wouldn't be offended and could simply close my version.

**Comment by LanieOkorodudu** [comment]
@LlGC-jop and @demiankatz , I have tested the fix, and it's working smoothly with both the Space bar and Enter key on my keyboard. Thank you @LlGC-jop and @demiankatz.

**Pull request merged**
